subsys/bluetooth/host/id.c

```diff
@@ -1254,6 +1344,151 @@ void bt_id_del(struct bt_keys *keys)
 		bt_le_ext_adv_foreach(adv_unpause_enabled, NULL);
 	}
 }
+
+static int conflict_check_and_replace(uint8_t id, struct bt_keys *keys)
+{
+	/* For the given key check if it has conflicts with other keys in the Resolving List
+	 * (such keys have BT_KEYS_ID_ADDED state and BT_KEYS_ID_CONFLICT flag set). If it does, we
+	 * need to remove the conflicting key from the Resolving List and add the new key.
+	 *
+	 * If the key is not in the Resolving List, we can add the new key right away.
+	 *
+	 * If advertiser for the conflicting key is enabled, we cannot remove the key from the
+	 * Resolving List, so we return an error.
+	 */
+
+	struct bt_keys *conflict;
+	struct bt_le_ext_adv *adv;
+
+	if (!(keys->flags & BT_KEYS_ID_CONFLICT)) {
+		LOG_DBG("Key has no conflicts for id %u addr %s", id,
+			bt_addr_le_str(&keys->addr));
+		return 0;
+	}
+
+	if (keys->state & BT_KEYS_ID_ADDED) {
+		LOG_DBG("Key is already added to resolving list for id %u addr %s",
+			id, bt_addr_le_str(&keys->addr));
+		return 0;
+	}
+
+	conflict = bt_id_find_conflict(keys, false);
+	if (conflict == NULL) {
+		/* bt_id_find_conflict returns only keys added to the Resolving List (state is
+		 * BT_KEYS_ID_ADDED). If the key has conflict, but no keys were added (for example,
+		 * if the last added key was removed after bt_unpair()), then this function will
+		 * return NULL.
+		 */
+		LOG_WRN("Conflicting key not in the Resolving List.");
+		goto add;
```

Comment:
If we flip to (conflict != NULL), we can remove the goto add; (should try to limit use of goto)